src/main/java/com/varukha/onlinebookstore/service/shoppingCart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,105 @@
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public void create(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto save(CreateCartItemRequestDto requestDto) {
+        CartItem existedCartItem = getCartItemByBookId(requestDto.getBookId());
+        if (existedCartItem == null || !existedCartItem.getBook().getId()
+                .equals(requestDto.getBookId())) {
+            return createNewCartItem(requestDto);
+        } else {
+            return updateCartItemQuantityWhenSave(existedCartItem, requestDto.getQuantity());
+        }
+    }
+
+    @Override
+    public CartItem getCartItemById(Long id) {
+        return cartItemRepository.findById(id).orElse(null);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCartByUserId(Long id) {
+        return shoppingCartRepository.findByUserId(getUserId())
+                .map(shoppingCartMapper::toDto)
+                .orElseThrow();
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto updateCartItemQuantity(Long id,
+                                              UpdateCartItemRequestDto cartItemRequestDto) {
+        CartItem cartItemById = getCartItemById(id);
+        int requestedQuantity = cartItemRequestDto.getQuantity();
+        cartItemById.setQuantity(requestedQuantity);
+        CartItem savedCartItem = cartItemRepository.save(cartItemById);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto updateCartItemQuantityWhenSave(CartItem existingCartItem, int quantity) {
+        existingCartItem.setQuantity(existingCartItem.getQuantity() + quantity);
+        CartItem savedCartItem = cartItemRepository.save(existingCartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto createNewCartItem(CreateCartItemRequestDto requestDto) {
+        CartItem cartItem = new CartItem();
+        cartItem.setShoppingCart(shoppingCartRepository
+                .findByUserId(getUserId())
+                .orElseThrow());
+        cartItem.setBook(bookRepository
+                .findById(requestDto.getBookId()).orElseThrow());
+        cartItem.setQuantity(requestDto.getQuantity());
+        CartItem savedCartItem = cartItemRepository.save(cartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private Long getUserId() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = (User) authentication.getPrincipal();
+        return user.getId();
+    }
```

Comment:
Maybe it's better to do it on controller level?

Reply:
No, I think it's correct to create one method and use it in all methods which need it. I think it's not a good idea to create utility method on controller level.